### PR TITLE
Cygwin build failed

### PR DESCRIPTION
**Pull request by chronoxor**

### CODE REVIEW

**Comment by acozzette** [comment]
@chronoxor I don't quite understand why defining this feature test macro fixes it. What version of GCC are you using? The readlink manpage on my machine says that readlink should be available in unistd.h if `_XOPEN_SOURCE >= 500 || _POSIX_C_SOURCE >= 200112L`, and it seems like that is true for recent GCC versions.

**Comment by chronoxor** [comment]
I signed it!

**Comment by chronoxor** [comment]
@acozzette, I use the most recent Cygwin 2.891 (64bit) with GCC compiler 6.4.0

The PR fixes only protobuf libs and compiler, but many tests are failed to build under Cygwin as well. So if you want to have a Cygwin build in Appveyour they have to be fixed too.

I have to say that first I found the similar issue in capnproto and this fix was accepted by Kenton Varda. More details you can find here - https://github.com/capnproto/capnproto/pull/660

**Comment by chronoxor** [comment]
#include <stdio.h>
#include <unistd.h>

int main(int argc, char* argv[])
{
    printf("%d\n", _XOPEN_SOURCE);
    printf("%d\n", _POSIX_C_SOURCE);
    return 0;
}

test.cpp:9:20: error: ‘_XOPEN_SOURCE’ was not declared in this scope
     printf("%d\n", _XOPEN_SOURCE);

printf("%d\n", _POSIX_C_SOURCE) outputs: 200809

**Comment by chronoxor** [comment]
In unistd.h -> sys/unistd.h readlink defined as:

#if __BSD_VISIBLE || __POSIX_VISIBLE >= 200112 || __XSI_VISIBLE >= 4
ssize_t readlink (const char *__restrict __path, char *__restrict __buf, size_t __buflen);
#endif

In simple test.cpp it is visible and
__BSD_VISIBLE = 1
__POSIX_VISIBLE = 200809
__XSI_VISIBLE = 0

I think something changed in protobuf includes/configurations. I use CMake to build protobuf under Cygwin with cmake -G "Unix Makefiles" .

**Comment by acozzette** [comment]
Could you try building with `make VERBOSE=1` and paste an example of the compiler command? Maybe that would tell us if somehow there's a problem with the compiler command line.

**Comment by chronoxor** [comment]
@acozzette, sure. Here is a cmake generate output:

 cmake -G "Unix Makefiles" .
-- The C compiler identification is GNU 6.4.0
-- The CXX compiler identification is GNU 6.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++.exe
-- Check for working CXX compiler: /usr/bin/c++.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - found
-- Found Threads: TRUE
-- Could NOT find ZLIB (missing:  ZLIB_LIBRARY ZLIB_INCLUDE_DIR)
-- Configuring done
-- Generating done
-- Build files have been written to: /home/chronoxor/protobuf/cmake

**Comment by chronoxor** [comment]
make VERBOSE=1 stops at the first compilation error (which could be fixed with _GNU_SOURCE as well):

[  0%] Building CXX object CMakeFiles/gmock.dir/home/chronoxor/protobuf/third_party/googletest/googlemock/src/gmock-all.cc.o
/usr/bin/c++.exe   -DGOOGLE_PROTOBUF_CMAKE_BUILD -DHAVE_PTHREAD -I/home/chronoxor/protobuf/cmake -I/home/chronoxor/protobuf/src -I/home/chronoxor/protobuf/third_party/googletest/googlemock -I/home/chronoxor/protobuf/third_party/googletest/googletest -I/home/chronoxor/protobuf/third_party/googletest/googletest/include -I/home/chronoxor/protobuf/third_party/googletest/googlemock/include  -std=c++11 -o CMakeFiles/gmock.dir/home/chronoxor/protobuf/third_party/googletest/googlemock/src/gmock-all.cc.o -c /home/chronoxor/protobuf/third_party/googletest/googlemock/src/gmock-all.cc
In file included from /home/chronoxor/protobuf/third_party/googletest/googletest/include/gtest/internal/gtest-linked_ptr.h:74:0,
                 from /home/chronoxor/protobuf/third_party/googletest/googlemock/include/gmock/internal/gmock-port.h:53,
                 from /home/chronoxor/protobuf/third_party/googletest/googlemock/include/gmock/internal/gmock-generated-internal-utils.h:44,
                 from /home/chronoxor/protobuf/third_party/googletest/googlemock/include/gmock/internal/gmock-internal-utils.h:45,
                 from /home/chronoxor/protobuf/third_party/googletest/googlemock/include/gmock/gmock-actions.h:46,
                 from /home/chronoxor/protobuf/third_party/googletest/googlemock/include/gmock/gmock.h:58,
                 from /home/chronoxor/protobuf/third_party/googletest/googlemock/src/gmock-all.cc:40:
/home/chronoxor/protobuf/third_party/googletest/googletest/include/gtest/internal/gtest-port.h: In function ‘int testing::internal::posix::FileNo(FILE*)’:
/home/chronoxor/protobuf/third_party/googletest/googletest/include/gtest/internal/gtest-port.h:2382:51: error: ‘fileno’ was not declared in this scope
 inline int FileNo(FILE* file) { return fileno(file); }
                                                   ^
/home/chronoxor/protobuf/third_party/googletest/googletest/include/gtest/internal/gtest-port.h: In function ‘char* testing::internal::posix::StrDup(const char*)’:
/home/chronoxor/protobuf/third_party/googletest/googletest/include/gtest/internal/gtest-port.h:2388:57: error: ‘strdup’ was not declared in this scope
 inline char* StrDup(const char* src) { return strdup(src); }
                                                         ^
/home/chronoxor/protobuf/third_party/googletest/googletest/include/gtest/internal/gtest-port.h: In function ‘FILE* testing::internal::posix::FDOpen(int, const char*)’:
/home/chronoxor/protobuf/third_party/googletest/googletest/include/gtest/internal/gtest-port.h:2416:71: error: ‘fdopen’ was not declared in this scope
 inline FILE* FDOpen(int fd, const char* mode) { return fdopen(fd, mode); }

**Comment by acozzette** [comment]
So it looks like the CMake build is not passing any special flags to the compiler other than `-std=c++11`. If you pass `-std=c++11` do you get the same values for macros (such as `__POSIX_VISIBLE`) that you already posted?

**Comment by chronoxor** [comment]
$ gcc -std=c++11 test1.cpp
__BSD_VISIBLE = 0
__POSIX_VISIBLE = 0
__XSI_VISIBLE = 0

$ gcc -std=c++11 test2.cpp
test.cpp: In function ‘int main(int, char**)’:
test.cpp:11:24: error: ‘readlink’ was not declared in this scope
     readlink ("", "", 0);

**Comment by chronoxor** [comment]
$ gcc -std=gnu++11 test1.cpp
__BSD_VISIBLE = 1
__POSIX_VISIBLE = 200809
__XSI_VISIBLE = 0

**Comment by xfxyjwf** [comment]
@chronoxor Can you build it with:
```
$ cmake -DCMAKE_CXX_FLAGS="-std=gnu++11" ...
```
?

Alternatively I think you can build protobuf using autotools.

**Comment by chronoxor** [comment]
/usr/bin/c++.exe ... std=gnu++11   -std=c++11 ...
/home/chronoxor/protobuf/third_party/googletest/googletest/include/gtest/internal/gtest-port.h: In function ‘int testing::internal::posix::FileNo(FILE*)’:
/home/chronoxor/protobuf/third_party/googletest/googletest/include/gtest/internal/gtest-port.h:2382:51: error: ‘fileno’ was not declared in this scope
 inline int FileNo(FILE* file) { return fileno(file); }
                                                   ^

**Comment by chronoxor** [comment]
It compiles if I comment out set(CMAKE_CXX_STANDARD 11) from protobuf/cmake/CMakeList.txt

**Comment by chronoxor** [comment]
Maybe we need something like this?

if (CYGWIN)
  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=gnu++11")
else()
  set(CMAKE_CXX_STANDARD 11)
endif()

**Comment by chronoxor** [comment]
CMAKE_CXX_EXTENSIONS - still have compilation errors. 

Ok, I'll prepare a new PR with corrected CMakeList.txt

File: cmake/CMakeLists.txt [hunk 1]
```diff
@@ -16,7 +16,11 @@ endif()
 project(protobuf C CXX)
 
 # Add c++11 flags
-set(CMAKE_CXX_STANDARD 11)
+if (CYGWIN)
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=gnu++11")
+else()
+  set(CMAKE_CXX_STANDARD 11)
+endif()
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
```
Comment (xfxyjwf) [inline cmake/CMakeLists.txt]: Can you move all of CMAKE_CXX_STANDARD/CMAKE_CXX_STANDARD_REQUIRED/CMAKE_CXX_EXTENSIONS inside the same else group?

Comment (chronoxor) [reply cmake/CMakeLists.txt]: Fixed!

**Comment by chronoxor** [comment]
@xfxyjwf, what do you think about adding Cygwin build to appveyor.yml?

**Comment by xfxyjwf** [comment]
@chronoxor Does appveyor support cygwin? Feel free to send us a pull request to add that.

**Pull request merged**